File: minimum_spanning_tree/prim_algorithm.py
# import libraries
import numpy as np
import sys

# function to test connected graph in kruskal's algorithm
def cycle_detection(vertices, mst_test):
    # prepaing variables for:
    matrix = np.copy(mst_test)  # copy of current mst from parameter
    visited = []                # visited vertices set
    cycle_maker = []            # vertices that make cycle
    flag = []                   # flag for each vertices (visited or no)

    # make all flag 0
    for i in range(vertices):
        flag.append(0)

    # cycle detection
    for i in range(vertices):
        # if current flag is 0 and visiting it, append vertex into visited set, and turn flag to 1
        if flag[i] == 0:
            visited.append(i)
            flag[i] = 1

        # if current flag is 1 and not in cycle_maker set, check for its neightbor
        if(flag[i] == 1) and (flag[i] not in cycle_maker):
            for j in range(vertices):
                # if matrix[i][j] (neighbor vertex) greater than 0 and its flag is 0
                # append neighbor vertex into visited set, turn matrix[i][j] and matrix[j][i] to 0
                # and set the neigbor flag to 1
                if(matrix[i][j] > 0) and (flag[j] == 0):
                    visited.append(j)
                    matrix[i][j] = matrix[j][i] = 0
                    flag[j] = 1
                # if matrix[i][j] (neighbor vertex) greater than 0 and its flag is 1
                # append the neighbor vertex into cycle_maker set
                elif(matrix[i][j] > 0) and (flag[j] == 1):
                    cycle_maker.append(j)
    
    # if there are some cycle maker, return true, else return false
    if cycle_maker != []:
        return True
    else:
        return False

# function to check connected graph
def connected_graph(matrix):
    # copy first adjacency matrix row to helper_matrix
    helper_matrix = np.copy(matrix[0, :])

    # do loop for each row
    for row in range(len(matrix)):
        # if current helper matrix not 0
        if helper_matrix[row] != 0:
            # do loop for each columnn
            for column in range(len(matrix)):
                # if current row and column greater than 0 and helper_matrix[column] still 0
                if(matrix[row][column] > 0) and (helper_matrix[column] == 0):
                    # change helper_matrix[column] to 1
                    helper_matrix[column] = 1

    # if 0 in helper matrix (graph not connected) return false, else return true
    if 0 in helper_matrix:
        return False
    else:
        return True


# preparing empty list for
matrix = []     # store adjacency matrix
mst = []        # mst adjacency matrix
edges = []      # list of weighted edges
mst_weight = 0  # store mst weight

# open matrix adjacency file and store in matrix list
matrix_file = open("graph.txt")
for row in matrix_file:
    matrix.append(row.split())

# count for vertices
vertices = len(matrix)

# convert string to int in matrix list
for i in range(vertices):
    matrix[i] = list(map(int, matrix[i]))
matrix = np.array(matrix)

# check for connected graph, exit if graph not connected
if not connected_graph(matrix):
    sys.exit("Graph tidak terhubung, proses tidak dapat dilanjutkan!")

# count for edges weight and its vertices and store in edges list
for i in range(vertices):
    for j in range(i, vertices):
        if matrix[i][j] != 0:
            edges.append({"from": i, "to": j, "weight": matrix[i][j]})

# make an empty mst adjacency matrix based on n x n elements
for i in range(vertices):
    mst.append([])
    for j in range(vertices):
        mst[i].append(0)

# print adjacency matrix
print("Matriks ketetanggan awal:")
print(matrix, end="\n\n")

# search first edge with minimum weight
print("Prim's MST Edges:")
# consider the first set in edge list is the minimum edge
minimal = edges[0]
# search for the minimum one, if there was replace the minimal before
for i in range(len(edges)):
    if minimal["weight"] > edges[i]["weight"]:
        minimal = edges[i]
# insert the vertices of minimum edge into mst
mst[minimal["from"]][minimal["to"]] = mst[minimal["to"]][minimal["from"]] = minimal["weight"]
print(f"{minimal['from']}-{minimal['to']} -> {minimal['weight']}")
# sum for the weight
mst_weight += minimal['weight']
# remove the minimal (selected) edge
edges.remove(minimal)

# next vertices
# prepare selected edge before
edge_before = []
while edges != []:
    # store the minimum before into edge before
    for vertex in ["from", "to"]:
        edge_before.append(minimal[vertex])

    # consider the first set in edge list is the minimum edge
    minimal = edges[0]
    # search for the actual minimum edges
    for i in range(len(edges)):
        if minimal["weight"] > edges[i]["weight"]:
            minimal = edges[i]
    
    # if the current minimal edge weight is in edge before
    if (minimal["from"] in edge_before) or (minimal["to"] in edge_before):
        # add the vertices from the minimal edge into mst
        mst[minimal["from"]][minimal["to"]] = mst[minimal["to"]][minimal["from"]] = minimal["weight"]

        # check for the cycle
        status = cycle_detection(vertices, mst)
        # if the selected minimum edge makes cycle
        if status:
            # make the current vertices 0 again, else print the vertices and sum weight
            mst[minimal["from"]][minimal["to"]] = mst[minimal["to"]][minimal["from"]] = 0
        else:
            print(f"{minimal['from']}-{minimal['to']} -> {minimal['weight']}")
            mst_weight += minimal['weight']
        
        # remove selected edge
        edges.remove(minimal)

# print the edge information
print("\nPrim's MST:")
mst = np.array(mst)
print(mst, end="\n\n")

print("Bobot graph:", mst_weight)

File: minimum_spanning_tree/prim_algorithm.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <math.h>
#include "graph.h"

using namespace std;

int main() {
    ifstream matrix_file;
    int vertices = 0;
    int buffer;
    int **matrix, **mst;
    int mst_weight = 0;
    vector<vector<int>> edges;
    vector<int> used_vertices;
    
    matrix_file.open("graph.txt");
    while(!matrix_file.eof()) {
        matrix_file >> buffer;
        vertices++;
    }
    matrix_file.close();
    vertices = sqrt(vertices);

    matrix = (int**)malloc(vertices * sizeof(*matrix));
    for(int i = 0; i < vertices; i++) {
        matrix[i] = (int*)malloc(vertices * sizeof(int));
    }

    matrix_file.open("graph.txt");
    for(int i = 0; i < vertices; i++) {
        for(int j = 0; j < vertices; j++) {
            matrix_file >> matrix[i][j];
        }
    }
    matrix_file.close();

    if(!connected_graph(vertices, matrix)) {
        cout << "Graph tidak terhubung, proses tidak dapat dilanjutkan." << endl;
        exit(0);
    }

    cout << "Matriks ketetanggaan awal:" << endl;
    for(int i = 0; i < vertices; i++) {
        for(int j = 0; j < vertices; j++) {
            cout << matrix[i][j] << " ";
        }
        cout << endl;
    }
    
    for(int i = 0; i < vertices; i++) {
        for(int j = i; j < vertices; j++) {
            if(matrix[i][j] != 0) {
                edges.push_back({i, j, matrix[i][j]});
            }
        }
    }

    mst = (int**)malloc(vertices * sizeof(*mst));
    for(int i = 0; i < vertices; i++) {
        mst[i] = (int*)malloc(vertices * sizeof(int));
    }

    cout << endl << "Prim's MST Edges:" << endl;
    int min_index = 0;
    for(int i = 0; i < edges.size(); i++) {
        if(edges[min_index][2] > edges[i][2]) {
            min_index = i;
        }
    }
    mst[edges[min_index][0]][edges[min_index][1]]
        = mst[edges[min_index][1]][edges[min_index][0]]
        = edges[min_index][2];

    mst_weight += edges[min_index][2];
    cout << edges[min_index][0] << "-" << edges[min_index][1];
        cout << " -> " << edges[min_index][2] << endl;
        
    used_vertices.push_back(edges[min_index][0]);
    used_vertices.push_back(edges[min_index][1]);
    edges.erase(edges.begin() + min_index);

    int weight_before;
    while(!edges.empty()) {
        weight_before = edges[min_index][2];
        min_index = 0;
        for(int i = 0; i < edges.size(); i++) {
            if((edges[min_index][2] > edges[i][2]) and (edges[i][2] > weight_before)) {
                min_index = i;
            }
        }
        
        if(neighbor_checking(used_vertices, edges[min_index][0])
            or neighbor_checking(used_vertices, edges[min_index][1])) {
                mst[edges[min_index][0]][edges[min_index][1]]
                    = mst[edges[min_index][1]][edges[min_index][0]]
                    = edges[min_index][2];

                if(cycle_detection(vertices, mst)) {
                    mst[edges[min_index][0]][edges[min_index][1]]
                            = mst[edges[min_index][1]][edges[min_index][0]]
                            = 0;
                } else {
                    mst_weight += edges[min_index][2];
                    cout << edges[min_index][0] << "-" << edges[min_index][1];
                    cout << " -> " << edges[min_index][2] << endl;

                    used_vertices.push_back(edges[min_index][0]);
                    used_vertices.push_back(edges[min_index][1]);
                }

                edges.erase(edges.begin() + min_index);
        }        
    }

    cout << endl << "Prim's MST:" << endl;
    for(int i = 0; i < vertices; i++) {
        for(int j = 0; j < vertices; j++) {
            cout << mst[i][j] << " ";
        }
        cout << endl;
    }

    cout << endl << "Prim's MST Weight: " << mst_weight << endl;

    return 0;
}